File: import-dataset.php
<?php

/**
 * @file import-dataset.php
 *
 */

$restart = FALSE;
$server = 'localhost:9200';
$url = 'http://www.bco-dmo.org/dataset/646115/data/download';
$index = 'ec-ocean-protein-portal';
$type = 'dataset_646115_v1';
$schema_url = 'https://raw.githubusercontent.com/ashepherd/elasticsearch-ocean-proteomics/master/dataset_646115_schema.json';
$newline = "\n";
$delim = "\t";
$base_http_opts = array(
  CURLOPT_RETURNTRANSFER => FALSE,
  CURLOPT_TIMEOUT => 10,
  CURLOPT_VERBOSE => TRUE,
);

$tsv = fopen($url, "r");
if ($tsv == FALSE) {
  echo "Could not open $url" . $newline;
  exit -1;
}

$columns = FALSE;
$length = 0;
$line_idx = 0;
while (!feof($tsv)) {
  $line = fgets($tsv);
  if (!$columns) {
    $columns = explode($delim, trim($line));
    $length = count($columns);

    if ($restart) {
      // Delete ElasticSearch Index.
      $delete_options = array(
        CURLOPT_CUSTOMREQUEST => 'DELETE',
      );
      $delete = _curl_request('http://' . $server . '/' . $index . '/_mapping/' . $type, $delete_options + $base_http_opts);
      echo "DELETE INDEX: " . $delete['response'] . $newline;
      // Create ElasticSearch Index.
      $schema = file_get_contents($schema_url);
      $insert_options = array(
        CURLOPT_CUSTOMREQUEST => 'PUT',
        CURLOPT_POSTFIELDS => $schema,
      );
      $insert = _curl_request('http://' . $server . '/' . $index . '/_mapping/' . $type, $insert_options + $base_http_opts);
      echo "PUT INDEX: " . $insert['response'] . $newline;
    }
  }
  else {
    $line_idx++;
    $values = explode($delim, trim($line));
    if ($length !== ($value_len = count($values))) {
      // Skip this row.
      echo $value_len . " >> " . $line . $newline;
    }
    else {
      // Add to ElasticSearch
      $doc = array('_data_table_row' => array('value' => $line_idx));
      $lat = FALSE;
      $lon = FALSE;
      foreach ($columns as $idx => $name) {
        $doc[$name]['value'] = $values[$idx];
        if ('lon' == $name) {
          $lon = $values[$idx];
        }
        elseif ('lat' == $name) {
          $lat = $values[$idx];
        }
      }
      if ($lon && $lat) {
        $doc['_coordinate'] = array('lat' => $lat, 'lon' => $lon);
      }
      $es_url = 'http://' . $server . '/' . $index . '/' . $type . '/' . $line_idx;
      $head_options = array(
        CURLOPT_HEADER => TRUE,
        CURLOPT_NOBODY => TRUE,
      );
      $head = _curl_request($es_url, $head_options + $base_http_opts);
      echo "CHECK " . $line_idx . ": " . $head['code'] . $newline;
      if (200 == $head['code']) {
        $update_options = array(
          CURLOPT_CUSTOMREQUEST => 'POST',
          CURLOPT_POSTFIELDS =>  '{"doc" : ' . json_encode($doc) . '}',
        );
        $update = _curl_request($es_url . '/_update', $update_options + $base_http_opts);
        if (200 == $update['code']) {
          echo "Updated row: $line_idx - " . $update['response'] . $newline;
        }
        else {
          echo "Error updating row: $line_idx - [" . $update['code'] . ']: '. $update['response'] . $newline;
          echo "  " . $update['log'] . $newline;
        }
      }
      else {
        $create_options = array(
          CURLOPT_CUSTOMREQUEST => 'PUT',
          CURLOPT_POSTFIELDS =>  json_encode($doc),
        );
        $create = _curl_request($es_url . '/_create', $create_options + $base_http_opts);

        if (201 == $create['code']) {
          echo "Created row: $line_idx - " . $create['response'] . $newline;
        }
        else {
          echo "Error creating row: $line_idx - [" . $create['code'] . ']: ' . $create['response'] . $newline;
          echo "  " . $create['log'] . $newline;
        }
      }
    }
  }
}

/**
 * Make an cURL HTTP Request.
 *
 * @param string $url
 *   The URL to request
 *
 * @param array $options
 *   THe PHP cURL options
 *
 * @return array
 *    'code' => the HTTP status code
 *    'response' => the response data
 */
function _curl_request($url, $opts) {

  ob_start();
  $out = fopen('php://output', 'w');

  $ch = curl_init($url);
  curl_setopt_array($ch, $opts);
  curl_setopt($ch, CURLOPT_STDERR, $out);
  $response = curl_exec($ch);
  $status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
  curl_close($ch);

  fclose($out);
  $log = ob_get_clean();

  return array(
    'code' => $status,
    'response' => $response,
    'log' => $log,
  );
}

File: import-dataset.py
import requests
import urllib2
import csv
import json
import getopt
import sys

def main(argv):

  rebuild_index = False
  server = 'http://localhost:9200'
  index = 'ec-ocean-protein-portal'
  es_type = 'dataset_646115_v1'

  help = 'import-dataset.py -r <rebuild ES index eg. true|false> -s <ES server eg. http://localhost:9200> -i <ES index name> -t <ES document type>'
  try:
    opts, args = getopt.getopt(argv,"hr:s:i:t:",["rebuild=","server=","index=","type="])
  except getopt.GetoptError:
    print help
    sys.exit(2)
  for opt, arg in opts:
    if opt == '-h':
       print help
       sys.exit()
    elif opt in ("-r", "--rebuild"):
       rebuild_index = arg.lower() == "true"
    elif opt in ("-s", "--server"):
       server = arg
    elif opt in ("-i", "--index"):
       index = arg
    elif opt in ("-t", "--type"):
       es_type =arg

  print 'ES Server: "', server
  print 'ES Index "', index
  print 'ES Doc Type "', es_type
  print 'Rebuild? "', rebuild_index

  index_settings = {'settings': {'index':{'number_of_shards': 1, 'number_of_replicas': 0}}}

  url = 'http://www.bco-dmo.org/dataset/646115/data/download'
  schema_url = 'https://raw.githubusercontent.com/ashepherd/elasticsearch-ocean-proteomics/master/dataset_646115_schema.json'

  delim = '\t'

  dataset = requests.get(url)
  tsv = dataset.text.split('\n')
  columns = []
  length = 0
  line_idx = 0

  for line in tsv:
    if not columns:
      columns = line.split(delim)
      length = len(columns)

      if rebuild_index:
        index_url = server + '/' + index
        mapping_url = index_url + '/_mapping/' + es_type
        r = requests.delete(mapping_url)
        print "DELETE INDEX: " + r.text

        r = requests.put(index_url, data = json.dumps(index_settings))

        schema = requests.get(schema_url)
        r = requests.put(mapping_url, data = schema.text)
        print r.status_code
        print "PUT SCHEMA: " + r.text

    else:
      line_idx += 1
      values = line.split(delim)
      values_len = len(values)
      if values_len != length:
        print 'Missing data on line ' + `(line_idx + 1)` + '. Found only ' + `values_len` + ' values, but expected ' + `length`
        print 'Line: "' + line + '"'
        exit()

      data = {}
      lat = False
      lon = False

      data['_data_table_row'] = {'value': line_idx}

      for idx, column_label in enumerate(columns):
        data[column_label] = {'value': values[idx]}
        if 'lon' == column_label:
          lon = values[idx]
        elif 'lat' == column_label:
          lat = values[idx]

      if (lat and lon):
        data['_coordinate'] = {'lat': lat, 'lon': lon}

      es_url = server + '/' + index + '/' + es_type + '/' + `line_idx`
      r = requests.head(es_url)
      if r.status_code == 200:
        # Update the row
        r = requests.post(es_url + '/_update', data = json.dumps({'doc': data}))
        if r.status_code == 200:
          print 'Updated row ' + `line_idx` + ': ' + r.text
        else:
          print 'ERROR updating row ' + `line_idx` + '[' + `r.status_code` + ']: ' + r.text

      else:
        r = requests.put(es_url + '/_create', data = json.dumps(data))
        if r.status_code == 201:
          print 'Created row ' + `line_idx` + ': ' + r.text
        else:
          print 'ERROR creating row ' + `line_idx` + '[' + `r.status_code` + ']: ' + r.text

if __name__ == "__main__":
   main(sys.argv[1:])
